language/libra-tools/writeset-transaction-generator/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use libra_types::{
+    account_address::AccountAddress,
+    transaction::{Script, Transaction},
+};
+use libra_types::{account_config::libra_root_address, transaction::WriteSetPayload};
+use std::io::Write;
+use stdlib::compile_script;
+use tempfile::NamedTempFile;
+
+fn compile_admin_script(input: &str) -> Result<Script> {
+    let mut temp_file = NamedTempFile::new()?;
+    temp_file.write_all(input.as_bytes())?;
+    let cur_path = temp_file.path().to_str().unwrap().to_owned();
+    Ok(Script::new(compile_script(cur_path), vec![], vec![]))
+}
+
+pub fn encode_remove_validators_transaction(validators: Vec<AccountAddress>) -> Transaction {
+    assert!(!validators.is_empty(), "Unexpected validator set length");
+    let mut script_body = "".to_string();
+    for address in validators {
+        script_body.push_str(
+            format!(
+                "LibraSystem::remove_validator(libra_root, 0x{:?});\n",
+                address
+            )
+            .as_str(),
+        );
+    }
+
+    let script = compile_admin_script(
+        format!(
+            "
+    script {{
+        use 0x1::LibraSystem;
+        fun main(libra_root: &signer) {{
+            {}
+        }}
+    }}",
+            script_body
+        )
+        .as_str(),
+    )
+    .expect("Cannot compile script");
+
+    Transaction::GenesisTransaction(WriteSetPayload::Script {
+        script,
+        execute_as: libra_root_address(),
+    })
+}
+
+pub fn encode_halt_network_transaction() -> Transaction {
+    let script = "
+    script {
+        use 0x1::LibraTransactionPublishingOption;
+        fun main(libra_root: &signer) {
+            LibraTransactionPublishingOption::set_open_script(libra_root);
+            LibraTransactionPublishingOption::add_to_script_allow_list(libra_root, x\"0000000000000000000000000000000000000000000000000000000000000000\");
+        }
+    }";
+
+    Transaction::GenesisTransaction(WriteSetPayload::Script {
```

Comment:
those expected to be signed by AOS should be RawTransaction instead of Transaction::GenesisTransaction? otherwise AOS has to unpack it first

also maybe we can add a libra version script too?

Reply:
There's a slight gotcha here. The admin script that needs to be signed by AOS will be slight different from the script here: The script here will contain one `signer` instead of two in regular AOS signed admin script. Thus the code here are intended to be only used by the GenesisTransaction use case.